### PR TITLE
various refactorings

### PR DESCRIPTION
**Pull request by dgkanatsios**
Various refactorings in the codebase.
Specifically:
- refactored expectations related code out of the GameServerBuild reconciler
- created a test for expectations related code
- refactored controller's initial configuration, using a struct to encapsulate all customizable options coming from environment variables
- split the "CI" pipeline into two, a "unit tests" and an "end to end tests" one
- various comment corrections here and there

### CODE REVIEW

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          go-version: 1.18.x
+      - name: build Docker images
+        run: make builddockerlocal
+      - name: install kind binaries
+        run: make installkind
+      - name: create kind cluster
+        run: make createkindcluster
+      - name: run end-to-end tests
+        run: make e2elocal
+      - name: Setup tmate session # instructions on https://github.com/marketplace/actions/debugging-with-tmate
```
Comment (javier-op) [inline .github/workflows/e2e.yml:44]: Is there a reason to setup the tmate session in the e2e tests but not in the unit tests?

Comment (dgkanatsios) [reply .github/workflows/e2e.yml:44]: End to end tests use a kind cluster which is easy to investigate using kubectl. Unit tests rely on K8s binaries (etcd, K8s api server) executing, so it’s trickier. We can revisit this later if needed.

File: pkg/operator/controllers/gameserver_expectations_test.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		e.addGameServerToUnderCreationMap(buildName, gsName1)
+		// make sure GameServerBuild and GameServer are in the underCreation map
+		gsUnderCreation, ok := e.gameServersUnderCreation.Load(buildName)
+		Expect(ok).To(BeTrue())
+		mmCreations := gsUnderCreation.(*MutexMap)
+		_, ok = mmCreations.data[gsName1]
+		Expect(ok).To(BeTrue())
+
+		// create a new GameServer and make sure it's added in the map
+		err = client.Create(context.TODO(), testGenerateGameServer(buildName, buildID, buildNamespace, gsName2))
```
Comment (javier-op) [inline pkg/operator/controllers/gameserver_expectations_test.go:38]: Usually context.TODO is used as a placeholder during development, context.Background is also an empty context intended for actual use.

Comment (dgkanatsios) [reply pkg/operator/controllers/gameserver_expectations_test.go]: refactored everywhere, thanks!

**Review by javier-op — APPROVED** [review]

**Pull request merged**